### PR TITLE
Make DataType and Reference generic

### PR DESCRIPTION
**Pull request by twogee**

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/91/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/85/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/93/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/87/

File: src/main/org/apache/tools/ant/taskdefs/optional/extension/ExtensionSet.java [hunk 1]
```diff
@@ -34,8 +34,7 @@
  *
  * @ant.datatype name="extension-set"
  */
-public class ExtensionSet
-    extends DataType {
+public class ExtensionSet extends DataType {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/taskdefs/optional/extension/ExtensionSet.java]: please keep whitespace changes separate from PRs (or better just don't create them at all)

File: src/main/org/apache/tools/ant/types/AbstractFileSet.java [hunk 2]
```diff
@@ -915,7 +920,7 @@ public String toString() {
     @Override
     public synchronized Object clone() {
         if (isReference()) {
-            return (getRef(getProject())).clone();
+            return getRef(getProject()).clone();
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/AbstractFileSet.java]: please keep style changes separate from PRs (or better just don't create them at all)

File: src/main/org/apache/tools/ant/types/ArchiveFileSet.java [hunk 3]
```diff
@@ -574,11 +570,8 @@ public int getDirMode() {
      * fileset if the project has been set).
      */
     private void checkArchiveAttributesAllowed() {
-        if (getProject() == null
-            || (isReference()
-                && (getRefid().getReferencedObject(
-                        getProject())
-                    instanceof ArchiveFileSet))) {
+        if (getProject() == null || (isReference()
+                && getRefid().getReferencedObject(getProject()) instanceof ArchiveFileSet)) {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/ArchiveFileSet.java]: please keep whitespace changes separate from PRs (or better just don't create them at all)

File: src/main/org/apache/tools/ant/types/DataType.java [hunk 4]
```diff
@@ -196,24 +196,44 @@ public static void pushAndInvokeCircularReferenceCheck(DataType dt,
...
     }
 
     /**
      * Performs the check for circular references and returns the
      * referenced object.
+     * @param <T> required reference type
+     * @param requiredClass the class that this reference should be a subclass of.
+     * @return the dereferenced object.
+     * @throws BuildException if the reference is invalid (circular ref, wrong class, etc).
+     * @since Ant 1.10
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/DataType.java]: needs the patch version

File: src/main/org/apache/tools/ant/types/FileList.java [hunk 5]
```diff
@@ -98,6 +98,13 @@ public File getDir(Project p) {
         return dir;
     }
 
+    /**
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/FileList.java]: please remove the unrelated new method

File: src/main/org/apache/tools/ant/types/FileList.java [hunk 6]
```diff
@@ -137,13 +144,10 @@ public void setFiles(String filenames) {
...
     /**
-     * Performs the check for circular references and returns the
-     * referenced FileList.
-     * @param p the current project
-     * @return the FileList represented by a referenced filelist.
+     * @return the list of files represented by this FileList.
      */
-    protected FileList getRef(Project p) {
-        return (FileList) getCheckedRef(p);
+    public String[] getFiles() {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/FileList.java]: please remove the unrelated new method

File: src/main/org/apache/tools/ant/types/FilterSet.java [hunk 7]
```diff
@@ -235,8 +235,8 @@ protected FilterSet(FilterSet filterset) {
      *
      * @return the filterset from the reference.
      */
-    protected FilterSet getRef() {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/FilterSet.java]: this is not backwards compatible

File: src/main/org/apache/tools/ant/types/Path.java [hunk 8]
```diff
@@ -721,7 +721,7 @@ public synchronized boolean isFilesystemOnly() {
      * @return the passed in ResourceCollection.
      */
     protected ResourceCollection assertFilesystemOnly(ResourceCollection rc) {
-        if (rc != null && !(rc.isFilesystemOnly())) {
+        if (rc != null && !rc.isFilesystemOnly()) {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/Path.java]: please keep style changes separate from PRs (or better just don't create them at all)

File: src/main/org/apache/tools/ant/types/PropertySet.java [hunk 9]
```diff
@@ -424,8 +424,8 @@ private void addPropertyNames(Set<String> names, Map<String, Object> props) {
      * referenced PropertySet.
      * @return the referenced PropertySet.
      */
-    protected PropertySet getRef() {
-        return getCheckedRef(PropertySet.class, "propertyset");
+    private PropertySet getRef() {
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/PropertySet.java]: this is not backwards compatible

File: src/main/org/apache/tools/ant/types/TarFileSet.java [hunk 10]
```diff
@@ -263,9 +273,7 @@ public Object clone() {
     private void checkTarFileSetAttributesAllowed() {
         if (getProject() == null
             || (isReference()
-                && (getRefid().getReferencedObject(
```
Comment (bodewig) [inline src/main/org/apache/tools/ant/types/TarFileSet.java]: please keep whitespace changes separate from PRs (or better just don't create them at all)

**Comment by bodewig** [comment]
Apart from my comments this looks good to me, thanks!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/96/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/90/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/97/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/91/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/98/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/92/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/99/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/93/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/100/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/94/

**Pull request closed without merging**